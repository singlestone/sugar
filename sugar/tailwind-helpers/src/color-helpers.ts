import type { CSSRuleObject, PluginAPI } from "tailwindcss/types/config";

import { SugarColorFamilies } from "./sugar-color-families.enum";
import { SugarPluginColorOptions } from "./sugar-plugin-color-options.interface";

export const convertHexToRGBA = (
  hex: string,
  alphaVariable: string | number | null = null
): string => {
  let r = 0;
  let g = 0;
  let b = 0;

  // 3 digits
  if (hex.length === 4) {
    r = +("0x" + hex[1] + hex[1]);
    g = +("0x" + hex[2] + hex[2]);
    b = +("0x" + hex[3] + hex[3]);

    // 6 digits
  } else if (hex.length === 7) {
    r = +("0x" + hex[1] + hex[2]);
    g = +("0x" + hex[3] + hex[4]);
    b = +("0x" + hex[5] + hex[6]);
  }

  return alphaVariable
    ? `rgba(${r}, ${g}, ${b}, ${alphaVariable})`
    : `rgb(${r}, ${g}, ${b})`;
};

export const getShadeValue = (
  theme: PluginAPI["theme"],
  color: string,
  shade: string | number,
  alphaVariable: string | number | null = null
) => {
  const colorValue: CSSRuleObject | string = theme(`colors.${color}`);
  let hex = "";
  if (colorValue) {
    hex =
      typeof colorValue === "string"
        ? colorValue
        : (colorValue[shade] as string);
  }
  return alphaVariable ? convertHexToRGBA(hex, alphaVariable) : hex;
};

export const getColor = (
  theme: PluginAPI["theme"],
  { colorFamily, value }: SugarPluginColorOptions = {}
): CSSRuleObject | string => {
  if (value) {
    return value;
  }
  const selector =
    colorFamily !== undefined ? colorFamily : SugarColorFamilies.ACCENT;
  return theme(selector);
};
