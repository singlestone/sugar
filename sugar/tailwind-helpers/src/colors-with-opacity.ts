const withOpacity =
  (
    variable: string,
    type: "hsl" | "rgb"
  ): ((
    arg: Partial<{ opacityVariable: string; opacityValue: number }>
  ) => string) =>
  ({ opacityValue = "" }) =>
    opacityValue
      ? `${type}(var(${variable}), ${opacityValue})`
      : `var(${variable})`;

export const hslWithOpacity = (variable: string) =>
  withOpacity(variable, "hsl");

export const rgbWithOpacity = (variable: string) =>
  withOpacity(variable, "rgb");
