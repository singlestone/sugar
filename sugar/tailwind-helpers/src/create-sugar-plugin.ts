import plugin from "tailwindcss/plugin";
import type { PluginAPI } from "tailwindcss/types/config";

import { defaultPluginConfig } from "./sugar-default-plugin-config";
import { SugarPluginOptions } from "./sugar-plugin-options.interface";

const defaultOptions = {
  prefix: "sugar",
};

export type SugarPluginComponent = (utils: PluginAPI, prefix: string) => void;

export type SugarTailwindPlugin = ReturnType<
  typeof plugin.withOptions<SugarPluginOptions>
>;

export const createSugarPlugin = (
  plugins: SugarPluginComponent[]
): SugarTailwindPlugin =>
  plugin.withOptions<SugarPluginOptions>(
    (options) => (utilFunctions: PluginAPI) => {
      const prefix: string = options?.prefix
        ? options.prefix
        : defaultOptions.prefix;
      plugins.forEach((sugarPlugin) => sugarPlugin(utilFunctions, prefix));
    },
    () => defaultPluginConfig
  );
