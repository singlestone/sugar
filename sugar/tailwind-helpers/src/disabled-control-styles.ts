import type { CSSRuleObject, PluginAPI } from "tailwindcss/types/config";
export const disabledControlStyles = (
  theme: PluginAPI["theme"]
): CSSRuleObject => ({
  cursor: "not-allowed",
  opacity: theme("opacity.50"),
});
