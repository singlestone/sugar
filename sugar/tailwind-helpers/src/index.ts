export * from "./color-helpers";
export * from "./colors-with-opacity";
export * from "./create-sugar-plugin";
export * from "./disabled-control-styles";
export * from "./focus-transition.styles";
export * from "./sugar-button-config.interface";
export * from "./sugar-color-families.enum";
export * from "./sugar-color-shades.enum";
export * from "./sugar-default-button-config";
export * from "./sugar-default-focus-config";
export * from "./sugar-default-form-config";
export * from "./sugar-default-measurements-config";
export * from "./sugar-default-plugin-config";
export * from "./sugar-measurements.enum";
export * from "./sugar-plugin-color-options.interface";
export * from "./sugar-plugin-options.interface";
