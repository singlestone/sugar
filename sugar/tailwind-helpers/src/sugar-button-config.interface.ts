export interface SugarButtonColorConfig {
  base: string;
  active: string;
  hover: string;
}

export interface SugarButtonGhostConfig {
  base: string;
}

export interface SugarButtonConfig {
  color: string;
  variants: {
    ghost: SugarButtonGhostConfig;
    outline: SugarButtonColorConfig;
    solid: SugarButtonColorConfig;
  }
}
