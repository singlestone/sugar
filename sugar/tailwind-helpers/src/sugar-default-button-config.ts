import { SugarButtonConfig } from "./sugar-button-config.interface";

export const sugarDefaultButtonConfig: { [key: string]: SugarButtonConfig } = {
  primary: {
    type: "filled",
    color: "accent",
    base: "600",
    active: "800",
    hover: "700",
  },
  secondary: {
    type: "outlined",
    color: "accent",
    base: "600",
    active: "200",
    hover: "100",
  },
  ghost: {
    type: "ghost",
    color: "accent",
    base: "600",
    active: "200",
    hover: "100",
  },
  destructive: {
    type: "filled",
    color: "destructive",
    base: "600",
    active: "800",
    hover: "700",
  },
  "secondary-destructive": {
    type: "outlined",
    color: "destructive",
    base: "600",
    active: "200",
    hover: "100",
  },
  "ghost-destructive": {
    type: "ghost",
    color: "destructive",
    base: "600",
    active: "200",
    hover: "100",
  },
  neutral: {
    type: "outlined",
    color: "neutral",
    base: "600",
    active: "200",
    hover: "100",
  },
  "ghost-neutral": {
    type: "ghost",
    color: "neutral",
    base: "600",
    active: "200",
    hover: "100",
  },
};
