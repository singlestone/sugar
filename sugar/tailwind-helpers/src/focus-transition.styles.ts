import type { CSSRuleObject, PluginAPI } from "tailwindcss/types/config";

import { getShadeValue } from "./color-helpers";
import { SugarColorShades } from "./sugar-color-shades.enum";

export interface FocusTransitionOptions {
  accentColors?: CSSRuleObject;
  initialRingWidth?: string;
  offsetColor?: string;
}

export interface FocusVisibleTransitionOptions {
  accentColors?: CSSRuleObject;
  offsetColor?: string;
}

interface TransitionConfig {
  initialRingWidth: string;
}

interface FocusVisibleTransitionConfig {
  accentColors: CSSRuleObject;
  offsetColor: string;
}

const getDefaultTransitionOptions = (): TransitionConfig => ({
  initialRingWidth: "30px",
});

const getDefaultFocusVisibleTransitionOptions = (
  theme: PluginAPI["theme"]
): FocusVisibleTransitionConfig => ({
  accentColors: theme("colors.blue"),
  offsetColor: theme("colors.white"),
});

export const focusTransition = (
  theme: PluginAPI["theme"],
  options?: FocusTransitionOptions
): CSSRuleObject => {
  const ftOptions = { ...getDefaultTransitionOptions(), ...options };
  return {
    "--tw-ring-inset": "var(--tw-empty,/*!*/ /*!*/)",
    "--tw-ring-offset-width": "2px",
    "--tw-ring-offset-color": "transparent",
    "--tw-ring-color": "transparent",
    "--sugar-ring-width": ftOptions.initialRingWidth,
    "--tw-ring-offset-shadow": `var(--tw-ring-inset) 0 0 0 var(--tw-ring-offset-width) var(--tw-ring-offset-color)`,
    "--tw-ring-shadow": `var(--tw-ring-inset) 0 0 0 calc(var(--sugar-ring-width) + var(--tw-ring-offset-width)) var(--tw-ring-color)`,
    boxShadow: `var(--tw-ring-offset-shadow), var(--tw-ring-shadow), var(--tw-shadow)`,
  };
};

export const focusVisibleTransition = (
  theme: PluginAPI["theme"],
  options?: FocusVisibleTransitionOptions
): CSSRuleObject => {
  const fvtOptions = {
    ...getDefaultFocusVisibleTransitionOptions(theme),
    ...options,
  };
  return {
    "--tw-ring-offset-color": fvtOptions.offsetColor,
    "--tw-ring-color": getShadeValue(
      fvtOptions.accentColors,
      SugarColorShades.Primary
    ),
    "--sugar-ring-width": theme("ringWidth.DEFAULT"),
    transitionProperty: theme("transitionProperty.DEFAULT"),
    transitionDuration: theme("transitionDuration.150"),
    transitionTimingFunction: theme("transitionTimingFunction.out"),
    outline: "none",
  };
};
