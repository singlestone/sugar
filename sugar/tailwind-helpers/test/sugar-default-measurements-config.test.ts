import { expect, test } from "vitest";

import { defaultPluginConfig, sugarDefaultMeasurementsConfig } from "../src";
import { createMockThemeFn } from "./mocks/create-mock-theme";

const theme = createMockThemeFn(
  {
    formControlHeight: "spacing.10",
    buttonSize: "spacing.5",
  },
  defaultPluginConfig
);

test("sugarDefaultFormConfig", () => {
  expect(sugarDefaultMeasurementsConfig(theme)).toMatchSnapshot();
});
