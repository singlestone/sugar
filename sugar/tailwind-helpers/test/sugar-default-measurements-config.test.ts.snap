// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sugarDefaultFormConfig 1`] = `
{
  "buttonSize": "spacing.5",
  "formControlHeight": "spacing.10",
}
`;
